### PR TITLE
Node ACL enforcement

### PR DESCRIPTION
**Pull request by schmichael**

### CODE REVIEW

File: nomad/node_endpoint.go [hunk 1]
```diff
@@ -536,6 +558,18 @@ func (n *Node) GetAllocs(args *structs.NodeSpecificRequest,
...
 	defer metrics.MeasureSince([]string{"nomad", "client", "get_allocs"}, time.Now())
 
+	// Check node read and namespace job read permissions
+	if aclObj, err := n.srv.resolveToken(args.SecretID); err != nil {
+		return err
+	} else if aclObj != nil {
+		if !aclObj.AllowNodeRead() {
+			return structs.ErrPermissionDenied
+		}
+		if !aclObj.AllowNsOp(args.RequestNamespace(), acl.NamespaceCapabilityReadJob) {
```
Comment (dadgar) [inline nomad/node_endpoint.go]: This one is a little more complex. You will have to filter the results of the iterator based on the namespace: https://github.com/hashicorp/nomad/pull/3222/files#diff-57d6a6426f4ff4ab5d9eecfbca4399daR584

Comment (dadgar) [reply nomad/node_endpoint.go]: And I think you should do it for each namespace on the alloc. You should add a map that caches the results per namespace.

File: nomad/node_endpoint_test.go [hunk 2]
```diff
@@ -656,6 +658,59 @@ func TestClientEndpoint_UpdateDrain(t *testing.T) {
...
+
+	// Update the status without a token and expect failure
+	dereg := &structs.NodeUpdateDrainRequest{
+		NodeID:       node.ID,
+		Drain:        true,
+		WriteRequest: structs.WriteRequest{Region: "global"},
+	}
+	{
+		var resp structs.NodeDrainUpdateResponse
+		assert.NotNil(msgpackrpc.CallWithCodec(codec, "Node.UpdateDrain", dereg, &resp), "RPC")
```
Comment (dadgar) [inline nomad/node_endpoint_test.go]: Assert permission denied

Comment (dadgar) [reply nomad/node_endpoint_test.go]: Here and in all others

File: nomad/node_endpoint.go (hunk 1 above)
Comment (schmichael) [reply nomad/node_endpoint.go]: > And I think you should do it for each namespace on the alloc.

I don't understand this part. Each alloc has exactly one Namespace right? So I should just filter out allocs where `alloc.Namespace != args.RequestNamespace()`?

**Pull request merged**